virt/driver.go

```diff
@@ -162,6 +175,27 @@ func (d *VirtDriverPlugin) SetConfig(cfg *base.Config) error {
 	d.virtualizer = v
 	d.taskGetter = v
 
+	// Generate a new libvirt connection for the network controller.
+	//
+	// TODO(jrasell): the compute and network should use the same connection,
+	// so we should refactor the connection setup in a future change.
+	libvirtConnection, err := libvirt.NewConnection(config.Emulator.URI, config.Emulator.User, config.Emulator.Password)
+	if err != nil {
+		return fmt.Errorf("virt: failed to create network libvirt connection: %w", err)
+	}
+
+	// Generate the new network controller and perform the initialization if
+	// this is required.
+	d.networkController = virtnet.NewController(d.logger, libvirt.NewConnect(libvirtConnection))
+
+	if !d.networkInit.Load() {
+		if err := d.networkController.Init(); err != nil {
+			return fmt.Errorf("virt: failed to init network controller: %w", err)
+		} else {
+			d.networkInit.Store(true)
+		}
+	}
```

Comment:
Using an atomic here implies we could be calling `SetConfig` multiple times or while the value is being read, in which case I'm not sure the atomic helps. Could this be a `sync.Once`?

Reply:
This is probably related to hashicorp/nomad#10839 which we should really fix someday :grimacing: 

I'm guessing James saw SetConfig getting called twice and used threadsafe patterns to manipulate its state, but (at least if 10839 is still accurate) SetConfig is only called _once per instance of the plugin_ so you don't have to worry about reentrancy of SetConfig.

Reply:
The main thing this tries to achieve is not to call the network init when it really doesn't need to. The init call itself is idempotent and all interaction with iptables uses host locking, so another process cannot attempt to also write configuration. Let me know if this changes your thoughts or views, but on the surface I am still happy with how this currently looks and works.

Reply:
So this is a "reduce unnecessary work" atomic, not a "for correctness" atomic. Works for me.